indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java

```diff
@@ -0,0 +1,694 @@
+    this.maxWaitTimeMillis = taskLockConfig.getBatchAllocationMaxWaitTime();
+    this.enabled = taskLockConfig.isBatchSegmentAllocation();
+
+    this.executor = executorFactory.create(1, "SegmentAllocQueue-%s");
+  }
+
+  @LifecycleStart
+  public void start()
+  {
+    log.info("Initializing segment allocation queue.");
+  }
+
+  @LifecycleStop
+  public void stop()
+  {
+    log.info("Tearing down segment allocation queue.");
+    executor.shutdownNow();
+  }
+
+  public boolean isEnabled()
+  {
+    return enabled;
+  }
+
+  private void scheduleQueuePoll(long delay)
+  {
+    executor.schedule(this::processBatchesDue, delay, TimeUnit.MILLISECONDS);
+  }
+
+  /**
+   * Gets the number of batches currently in the queue.
+   */
+  public int size()
+  {
+    return processingQueue.size();
+  }
+
+  /**
+   * Queues a SegmentAllocateRequest. The returned future may complete successfully
+   * with a non-null value or with a non-null value.
+   */
+  public Future<SegmentIdWithShardSpec> add(SegmentAllocateRequest request)
+  {
+    if (!isLeader.get()) {
+      throw new ISE("Cannot allocate segment if not leader.");
+    } else if (!isEnabled()) {
+      throw new ISE("Batched segment allocation is disabled.");
+    }
+
+    final AllocateRequestKey requestKey = new AllocateRequestKey(request, false);
+    final AtomicReference<Future<SegmentIdWithShardSpec>> requestFuture = new AtomicReference<>();
+
+    // Possible race condition:
+    // t1 -> new batch is added to queue or batch already exists in queue
+    // t2 -> executor pops batch, processes all requests in it
+    // t1 -> new request is added to dangling batch and is never picked up
+    // Solution: For existing batch, call keyToBatch.remove() on the key to
+    // wait on keyToBatch.compute() to finish before proceeding with processBatch().
+    // For new batch, keyToBatch.remove() would not wait as key is not in map yet
+    // but a new batch is unlikely to be due immediately, so it won't get popped right away.
+    keyToBatch.compute(requestKey, (key, existingBatch) -> {
+      if (existingBatch == null) {
+        AllocateRequestBatch newBatch = new AllocateRequestBatch(key);
+        requestFuture.set(newBatch.add(request));
+        return addBatchToQueue(newBatch) ? newBatch : null;
+      } else {
+        requestFuture.set(existingBatch.add(request));
+        return existingBatch;
+      }
+    });
+
+    return requestFuture.get();
+  }
+
+  /**
+   * Tries to add the given batch to the processing queue. If the queue is full,
+   * marks the batch as completed failing all the pending requests.
+   */
+  private boolean addBatchToQueue(AllocateRequestBatch batch)
+  {
+    batch.resetQueueTime();
+    if (processingQueue.offer(batch)) {
+      log.debug("Added a new batch [%s] to queue.", batch.key);
+      return true;
+    } else {
+      log.warn("Cannot add batch [%s] as queue is full. Failing [%d] requests.", batch.key, batch.size());
```

Comment:
what happens as a result of this? I assume it will be caused when the metadata store is slow. can there be any other reason? should we add some possible corrective action here like checking metadata store sizing?

Reply:
It can also happen when things keep getting retried, due to frequent updates to the set of used segments. It would be up to the client to retry in such cases.

The action can be to look at the emitted metrics `batch/runTime` and `batch/attempts`. A greater value of the first would be caused by metadata slowness, the second is retries. The user can then determine if the metadata store needs resizing. What do you think?